Push `codepoints()` implementation down into encoded variants

choznerol opened the pull request:
Fix #1728

This PR moves the encoding-specific code from `codepoints.rs` to encoded variants.

- The original version [consumes](https://github.com/artichoke/artichoke/compare/choznerol/1728-Implement-a-Codepoints-iterator-for-each-encoded-string-variant-in-spinoso-string?expand=1#diff-a2d2fa818d5b61fb277dd19f5a3a58b696944224b071fe8faee5bca3c19103d3L1923) the `String` to produce a `Codepoints` iterator, while the new version doesn't consume the `String` but allocate a `Vec` to create iterator. I'm not sure if the previous approach is preferred, if so I can try to refactor it.
- I made 589977d and 24837dd separate commits, as they may or may not be the preferred approach.

choznerol committed:
Push codepoints implementation down into encoded variants

Fix #1728

choznerol committed:
Fix clippy::unnessary_wraps

choznerol committed:
Prefer 'char' over 'u32' to represent codepoint

lopopolo commented on spinoso-string/src/enc/ascii/mod.rs:
```diff
@@ -280,6 +281,18 @@ impl AsciiString {
     }
+
+    #[inline]
+    pub fn codepoints(&self) -> Codepoints {
+        let iter = self
+            .inner
+            .as_slice()
+            .bytes()
+            .map(char::from)
+            .collect::<Vec<_>>()
+            .into_iter();
+        Codepoints::from(iter)
+    }
```
usually I'd implement this by having each encoding have their own ` CodePoints` iterator.

Then the codepoints iterator for string would be defined like this:

```rust
pub struct Codepoints<'a> {
    state: State<'a>,
}

enum State<'a> {
    Ascii(ascii::Codepoints<'a>),
    Binary(binary::Codepoints<'a>),
    Utf8(utf8::Codepoints<'a>),
}
```

Take a look at how `Inspect` is implemented:

https://github.com/artichoke/artichoke/blob/0ae1fb15dee52dc736d05302804625a4d4c8819c/spinoso-string/src/enc/inspect.rs#L5-L35

This avoids the allocation to normalize everything to `IntoIter`.

choznerol replied:
Thanks! This was exactly what I tried to figure out but couldn't. I reimplement with this design in b8729b7.

lopopolo replied:
awesome! new revision looks great. I kinda love this pattern of handling the various encodings and Rust's lifetime support makes the code super easy to reason about (and elegant IMO).

lopopolo commented on spinoso-string/src/enc/ascii/mod.rs:
```diff
@@ -280,6 +281,18 @@ impl AsciiString {
+            .inner
+            .as_slice()
+            .bytes()
+            .map(char::from)
```
I'm not sure it is correct to call `char::from`. What is the behavior when the encoding is not valid?

```console
$ irb
[3.2.2] > s = "\0\xFF"
=> "\u0000\xFF"
[3.2.2] > s.codepoints
(irb):2:in `codepoints': invalid byte sequence in UTF-8 (ArgumentError)
        from (irb):2:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
[3.2.2] > s.dup.b.codepoints
=> [0, 255]
[3.2.2] > s.dup.force_encoding('ascii').codepoints
=> [0, 255]
```

choznerol replied:
I reverted to use `u32::from` instead of `char::from`.

Also tested this for ASCII in e6be81d and asserted invalid bytes are just interpreted as bytes (e.g. `\xFF` -> `255`) instead of raising error:

https://github.com/artichoke/artichoke/blob/e6be81d1dbd798b8215f7d4031f83611b6ff9b8b/spinoso-string/src/enc/ascii/codepoints.rs#L52-L56

However, from `airb` an invalid UTF-8 code point error is still raised even with `force_encoding('ascii')`, as of e6be81d1dbd798b8215f7d4031f83611b6ff9b8b:
```console
$ cargo run --bin airb
    Finished dev [unoptimized + debuginfo] target(s) in 0.19s
     Running `target/debug/airb`
artichoke 0.1.0-pre.0 (2023-10-20 revision 7021) [aarch64-apple-darwin]
[rustc 1.70.0 (90c541806 2023-05-31) on aarch64-apple-darwin]
>>> "hello".codepoints
=> [104, 101, 108, 108, 111]
>>> "hello💎".codepoints
=> [104, 101, 108, 108, 111, 128142]
>>>
>>> s = "\0\xFF"
=> "\x00\xFF"
>>> s.codepoints
Traceback (most recent call last):
	2: from (airb):1
	1: from (airb):1:in codepoints
ArgumentError (invalid byte sequence in UTF-8)
>>> s.dup.b.codepoints
=> [0, 255]
>>> s.dup.force_encoding('ascii').codepoints
Traceback (most recent call last):
	2: from (airb):1
	1: from (airb):1:in codepoints
ArgumentError (invalid byte sequence in UTF-8)
>>>
```

Since `map(u32::from)` is introduced before this PR, perhaps that's related to another issue #1726?

lopopolo replied:
ahh yea, `String#force_encoding` is a stub right now, not properly implemented:

https://github.com/artichoke/artichoke/blob/f32bd63872b8b5b247c3bcb6516801767290057f/artichoke-backend/src/extn/core/string/string.rb#L519-L526

I think getting that fixed is part of @b-n's work on implementing `Encoding`

lopopolo commented on spinoso-string/src/enc/utf8/owned.rs:
```diff
@@ -146,6 +147,16 @@ impl Utf8String {
     }
 
+    #[inline]
+    pub fn codepoints(&self) -> Result<Codepoints, CodepointsError> {
```
this should be implemented on the UTF-8 borrowed type. It does not need to mutate the backing buffer.

choznerol replied:
I implemented this on `Utf8Str` instead in b8729b7.

lopopolo commented on spinoso-string/src/enc/utf8/owned.rs:
```diff
@@ -146,6 +147,16 @@ impl Utf8String {
 
+    #[inline]
+    pub fn codepoints(&self) -> Result<Codepoints, CodepointsError> {
+        if let Ok(s) = self.inner.as_slice().to_str() {
```
this should use `self.is_valid_encoding()` which is SIMD-optimized.

choznerol replied:
I end up using a single `simdutf8::basic::from_utf8()` in b8729b7.

lopopolo commented on spinoso-string/src/lib.rs:
```diff
@@ -1919,7 +1919,7 @@ impl String {
     /// [`str::from_utf8`]: core::str::from_utf8
     #[inline]
-    pub fn codepoints(&self) -> Result<Codepoints<'_>, CodepointsError> {
+    pub fn codepoints(&self) -> Result<Codepoints, CodepointsError> {
```
this should ideally not allocate.

choznerol replied:
The b8729b7 version doesn't use `alloc::vec` anymore 👍

choznerol commented:
Just found out I somehow missed the notification 😓  Will get back to this soon.

choznerol committed:
Revert "Prefer 'char' over 'u32' to represent codepoint"

This reverts commit 24837ddc777a9e80501f9fc8f0e97f08458ead10.

choznerol committed:
Revert "Fix clippy::unnessary_wraps"

This reverts commit 589977dd66ae394a930abff2380299c6b087a9dd.

choznerol committed:
Revert "Push codepoints implementation down into encoded variants"

This reverts commit 708ff943bb22e77c6c7dd99f031d579c62f7cc9e.

choznerol committed:
Push codepoints implementation down into encoded variants

Reimplement #2637 according to review commetn https://github.com/artichoke/artichoke/pull/2637#pullrequestreview-1608342189
This differs from the initial version in:
1. For utf-8 variant, impelment codepoint for the borrowed type (Utf8Str) instead of the owned type, so allocation is avoided.
2. Use 'simdutf8::basic::from_utf8' for performance
3. Give each encoding their own `Codepoints` iterator (mostly reference String::inspect)

choznerol committed:
Test Codepoints of encoded variants

lopopolo commented on spinoso-string/src/enc/utf8/borrowed/codepoints.rs:
```diff
@@ -0,0 +1,57 @@
+    fn try_from(s: &'a Utf8Str) -> Result<Self, Self::Error> {
+        match simdutf8::basic::from_utf8(s.as_bytes()) {
+            Ok(s) => Ok(Self { inner: s.chars() }),
+            Err(_) => Err(CodepointsError::invalid_utf8_codepoint()),
```
Can we add this in a comment as documentation?

```
[3.2.2] > s = "abc\xFFxyz"
=> "abc\xFFxyz"
[3.2.2] > s.encoding
=> #<Encoding:UTF-8>
[3.2.2] > s.codepoints
(irb):5:in `codepoints': invalid byte sequence in UTF-8 (ArgumentError)
        from (irb):5:in `<main>'
        from /usr/local/var/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/irb-1.6.2/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.2.2/bin/irb:25:in `<main>'
```

lopopolo replied:
I pushed a followup commit with this: 131fd3dab4f2451e7393f626e088c189b61f9732

lopopolo committed:
Add docs for argument error on invalid encoding in String#codepoints

lopopolo commented:
So sorry it took me so long to get back to you on this @choznerol! Once CI is green this is going in. Thanks for addressing all of my feedback (and for the ping by re-requesting my review 😅)

choznerol commented:
> So sorry it took me so long to get back to you on this @choznerol! Once CI is green this is going in. Thanks for addressing all of my feedback (and for the ping by re-requesting my review 😅)

Not at all. Thank you for the detailed code review!
